feat: Add custom Last.fm API and ListenBrainz API support to Audioscrobbler

vysmaty opened the pull request:
## Summary
This PR adds ListenBrainz API support to the AudioScrobbler plugin and renames it from "Last.fm AudioScrobbler" to simply "AudioScrobbler" to better reflect its expanded capabilities. The plugin now supports multiple scrobbling services while maintaining full backward compatibility.

## Features Added
- **ListenBrainz API Integration**: Complete implementation of ListenBrainz scrobbling protocol
- **ListenBrainz Now Playing Support**: Real-time track notifications with periodic updates every 30 seconds
- **Flexible Configuration**: Support for custom API URLs enabling compatibility with various services
- **Enhanced UI**: Updated configuration interface with examples and clearer instructions
- **Plugin Rename**: Changed from "Last.fm AudioScrobbler" to "AudioScrobbler" to reflect broader scope

## UI Changes
- **Scrobbling service selection**: Last.fm (v1.2) / ListenBrainz
- **Optional API Base URL**: Custom API URL field is optional - if left empty, defaults are used (post.audioscrobbler.com for Last.fm, api.listenbrainz.org for ListenBrainz)
- **Enhanced Help Description**: Added examples for various custom services:
  - Maloja (Last.fm v1.2 API): `http://localhost:42010/apis/audioscrobbler_legacy`
  - Maloja (ListenBrainz API): `https://maloja.yourdomain.com/apis/listenbrainz`
  - Koito (ListenBrainz API): `https://koito.yourdomain.com/apis/listenbrainz`
  - Multi-Scrobbler (ListenBrainz API): `https://scrobble.yourdomain.com/1/submit-listens`

<img width="1193" height="701" alt="image" src="https://github.com/user-attachments/assets/146cb15b-9313-4c52-8a39-35fb99d29916" />

  

<img width="1653" height="776" alt="image" src="https://github.com/user-attachments/assets/eef0f51f-b5fc-4619-a05b-07bf73a32801" />


## Tested Services
Successfully tested with multiple scrobbling services:
- **Last.fm** (existing functionality maintained)
- **ListenBrainz** (official instance)
- **Maloja** (both Last.fm API and ListenBrainz API endpoints)
- **Koito** (ListenBrainz API)
- **Multi-Scrobbler** (ListenBrainz API)

## Impact
This enhancement dramatically expands scrobbling possibilities for Lyrion users, allowing them to choose from a wide variety of music tracking services while maintaining full backward compatibility with existing Last.fm setups.

## Technical Details
- **Zero Breaking Changes**: Original Last.fm implementation (legacy API) remains mosty unchanged
- **Backward Compatibility**: Existing Last.fm accounts continue to work exactly as before
- **Account Type Detection**: Uses `api_type` field to distinguish between Last.fm and ListenBrainz protocols
- **ListenBrainz: Individual Item Submission**: Implements per-item submission to avoid API error in single mode
- **ListenBrainz: Proper Timer Management**: Cleanup of Now Playing timers when tracks change
- **Flexible API URLs**: Supports both default APIs and custom instances

## Translation Notes
- Updated and added translations
- Translation improvements for other languages would be welcome

## Note
- **This is a humble contribution to my favorite Lyrion project. The code would benefit from review by more experienced developers to ensure it meets the project's quality standards and best practices.**

## Breaking Changes
None - this is a backward-compatible enhancement that only adds new functionality.

michaelherger commented:
Thanks a lot! That sounds like a great enhancement! One which I always thought I should look into at some point myself 😁. 

Would you mind posting a screenshot in English? And how strict may I be before you get annoyed with my comments? 😂 

One thing I was wondering: right now the code often checks whether the type for an entry was defined, falling back to Last.fm if not. I assume this is to ease transition of existing configurations?  What you could also do is bump the prefs file version and have some migration code set the missing configuration parameters for existing entries. See eg. https://github.com/LMS-Community/slimserver/blob/public/9.1/Slim/Plugin/DateTime/Plugin.pm#L20-L39

And please try to sign-off the PR. See https://github.com/apps/dco. Thanks!

vysmaty committed:
Add ListenBrainz API support and rename plugin to AudioScrobbler

- Add comprehensive ListenBrainz API integration with JSON payload support
- Implement Now Playing notifications with periodic updates for ListenBrainz
- Add support for custom API URLs for various scrobbling services
- Rename plugin from 'Last.fm AudioScrobbler' to 'AudioScrobbler'
- Update UI with examples for Maloja, Koito, Multi-Scrobbler services
- Maintain full backward compatibility with existing Last.fm functionality
- Add Czech translations and improve help text
- Test coverage: Last.fm, ListenBrainz, Maloja, Koito, Multi-Scrobbler

Signed-off-by: vysmaty <10959584+vysmaty@users.noreply.github.com>
Signed-off-by: Martin Šebek <10959584+vysmaty@users.noreply.github.com>

vysmaty committed:
Add ListenBrainz to Plugin description.

Signed-off-by: vysmaty <10959584+vysmaty@users.noreply.github.com>
Signed-off-by: Martin Šebek <10959584+vysmaty@users.noreply.github.com>

vysmaty commented:
I've edited the screenshot. Tried the sign-off again. It's already signed twice, so I hope it's okay. 😄

Yes exactly, I wouldn't have thought of migrating existing accounts like that. I was trying not to break anything at the beginning, so there will be a surplus of safety fuses.

At first I was just adding the option to set the Last.fm url for the API to Maloja... 😂 And then it happened to work pretty quickly for ListenBrainz as well. With the Last.fm one, I personally had a problem with it fighting with https from the reverse proxy on the Maloja side... and it was still responding over http…

[ListenBrainz documentation](https://listenbrainz.readthedocs.io/en/latest/users/api-usage.html) was sometimes not clear to me in details, but it seemed to work on all four tested targets. Koito is more sensitive to string vs number… The multi-scrobber in turn needs that Now Playing.

I'll leave it up to you how strict you want to be, but I certainly don't want to exhaust you and myself in educating me in perl or coding basics. Really, it's a bit of dabbling on my part. Hopefully done responsibly and clearly on my part, but for example your link to [https://github.com/LMS-Community/slimserver/blob/public/9.1/Slim/Plugin/DateTime/Plugin.pm#L20-L39](https://github.com/LMS-Community/slimserver/blob/public/9.1/Slim/Plugin/DateTime/Plugin.pm#L20-L39) I understand, I'm interested in, but wheres is stored information about the version from which it will be migrated is not very clear me. And I don't want to break anything unnecessarily in my setup by trial and error. 😇

michaelherger commented on Slim/Plugin/AudioScrobbler/strings.txt:
```diff
@@ -63,84 +64,148 @@ SETUP_PLUGIN_AUDIOSCROBBLER_ACCOUNTS_DELETE
-	PL	Prześlij informacje o wszystkich odtwarzanych utworach do usługi Last.fm.
-	RU	Передавать (скробблить) на Last.fm все воспроизводимые дорожки.
-	SV	Ja, rapportera spelade spår till Last.fm (skrobbla).
+	CS	Odeslat (scroblovat) všechny přehrané skladby do Last.fm / ListenBrainz.
+	DA	Indsend (scrobble) oplysninger om alle afspillede numre til Last.fm / ListenBrainz
+	DE	Alle wiedergegebenen Titel an Last.fm / ListenBrainz melden (scrobbeln).
+	EN	Submit (scrobble) all tracks played to Last.fm / ListenBrainz.
+	ES	Enviar (hacer "scrobbling") todas las pistas reproducidas a Last.fm / ListenBrainz.
+	FI	Lähetä tiedot kaikista soitetuista kappaleista Last.fm / ListenBrainz:ään.
+	FR	Envoyer tous les morceaux lus vers Last.fm / ListenBrainz.
+	HU	Az összes lejátszott szám elküldése (scrobble) a Last.fm / ListenBrainz-re.
+	IT	Invia tutti i brani riprodotti in Last.fm / ListenBrainz.
+	NL	Alle gespeelde nummers naar Last.fm / ListenBrainz versturen (scrobble).
+	NO	Send (scrobble) alle spor som spilles av til Last.fm / ListenBrainz.
+	PL	Prześlij informacje o wszystkich odtwarzanych utworach do usługi Last.fm / ListenBrainz.
+	RU	Передавать (скробблить) на Last.fm / ListenBrainz все воспроизводимые дорожки.
+	SV	Ja, rapportera spelade spår till Last.fm / ListenBrainz (skrobbla).
```
Remove the product names, it's getting too busy, requiring the title to wrap.

michaelherger commented on Slim/Plugin/AudioScrobbler/HTML/EN/plugins/AudioScrobbler/settings/basic.html:
```diff
@@ -9,9 +9,24 @@
-		&nbsp;&nbsp;
-		<input type="password" class="stdedit" name="pref_password" id="password" value="" size="20">
+		<div style="margin-bottom:16px;">
+			<label for="username" style="margin-right:4px;">[% "SETUP_PLUGIN_AUDIOSCROBBLER_ACCOUNTS_USERNAME" | string %]</label>
+			<input type="text" class="stdedit" name="pref_username" id="username" value="" size="20">
+			&nbsp;&nbsp;
+			<label for="password" style="margin-right:4px;">[% "SETUP_PLUGIN_AUDIOSCROBBLER_ACCOUNTS_PASSWORD" | string %]</label>
+			<input type="password" class="stdedit" name="pref_password" id="password" value="" size="20">
```
Try to wrap each label/input combination into a single `span` or something which then can be marked not to wrap inside. Right now these look a bit confusing on small screens (mobile) as the input field would be separated from its label.
<img width="472" height="216" alt="Bildschirmfoto 2025-08-19 um 06 34 53" src="https://github.com/user-attachments/assets/e7749005-5723-4bb8-ada1-3a71831647e8" />

michaelherger commented on Slim/Plugin/AudioScrobbler/HTML/EN/plugins/AudioScrobbler/settings/basic.html:
```diff
@@ -9,9 +9,24 @@
+		</div>
+		<div class="prefDesc">[% "SETUP_PLUGIN_AUDIOSCROBBLER_ACCOUNTS_API_TYPE" | string %]</div>
+		<div>
+			<label for="api_type" style="margin-right:4px;">API:</label>
+			<select class="stdedit" name="pref_api_type" id="api_type">
+				<option [% IF (prefs.pref_api_type || '') == 'lastfm' %]selected="selected"[% END %] value="lastfm">Last.fm (v1.2)</option>
+				<option [% IF (prefs.pref_api_type || '') == 'listenbrainz' %]selected="selected"[% END %] value="listenbrainz">ListenBrainz API (Experimental)</option>
+			</select>
+			&nbsp;&nbsp;
+			<label for="api_url" style="margin-right:4px;">[% "SETUP_PLUGIN_AUDIOSCROBBLER_ACCOUNTS_CUSTOM_URL" | string %]</label>
+			<input type="text" class="stdedit" name="pref_api_url" id="api_url" value="[% prefs.pref_api_url || '' | html %]" size="20">
```
same as above: try to group label/input so they wrap together

michaelherger commented on Slim/Plugin/AudioScrobbler/Plugin.pm:
```diff
@@ -668,6 +709,191 @@ sub _submitNowPlayingError {
 	);
 }
 
+sub submitNowPlayingListenBrainz {
```
Instead of having all these `sub doSomethingListenBrainz`, please create a new modules à la `Slim::Plugin::AudioScrobbler::ListenBrainz` for each service or interface in which you have methods of the same name for each service. Then you'll only have to define which module to use as `my $serviceModule = 'Slim::Plugin::AudioScrobbler::ListenBrainz;` (or `...::LastFM`, or...), then `$serviceModule->submitNowPlaying()`.

If structured well you could have one implementation (LastFM) be the base class, then have the others inherit from that, only overwriting what needs to be overwritten.

michaelherger replied:
I don't know your level of Perl. Please let me know if you want to have some help with this part of it. But you've done great so far 😉.

vysmaty replied:
It's really beyond me here. I understood the intent. But I can't rewrite and divide it into modules like this without a lot of suffering. I've never written anything in Perl, if it weren't for the newfangled helpers (AI) that explain things here and there, I wouldn't even try adding ListenBrainz. 🫨

My second problem, is the different logic between Last.fm (legacy) and ListenBrainz, which I would have to somehow not screw up. They work significantly differently for me. One does the handshake and sends nowplaying about once. The other doesn't need the handshake and needs the nowplaying information repeatedly. 🤯

Let me know what you do with this. I'd be happy to test the functionality and debug for different services.  😇

michaelherger replied:
You and your friend AI have done an excellent job! I can probably look into this splitting out by service if you give me a day or two. 

Every now and then I'm amazed at what AI can do... what tool did you use?

vysmaty replied:
You're great. I will look forward to seeing what the outcome of the splitting looks like.

I first tried what Github Copilot can do with GPT 4.1 in VSCode and when I ran out of limit, it was Cline with Sonnet 4 that helped. Github Copilot is a good helper. Cline tries to do everything by itself, too much for my taste. A lot of care one has to take to still be an author and in control. I tried to keep it in the spirit of the original plugin and not have too much ballast.

Oh, and I guess next time I'll be squinting for Gemini CLI. :-)

michaelherger commented on Slim/Plugin/AudioScrobbler/Plugin.pm:
```diff
@@ -812,22 +1038,46 @@ sub submitScrobble {
-		$log->debug( 'Scrobbling ' . scalar( @{$queue} ) . ' queued item(s)' );
-		#$log->debug( Data::Dump::dump($queue) );
-	}
+       # Find out the API type and call ListenBrainz if necessary
+       my $accounts = getAccounts($client);
+       my ($api_type, $api_url, $api_key);
+       for my $acc (@{$accounts}) {
+	       if ($acc->{username} eq $account) {
+		       $api_type = $acc->{api_type} || 'lastfm';
+		       $api_url  = $acc->{api_url}  || '';
+		       $api_key  = $acc->{password} || '';
+		       last;
+	       }
+       }
+       if ( main::DEBUGLOG && $log->is_debug ) {
```
Please make sure you always use tabs rather than spaces for indentation. Everywhere.

vysmaty replied:
I hope I found all the occurrences. :-)

michaelherger commented on Slim/Plugin/AudioScrobbler/Settings.pm:
```diff
@@ -61,68 +61,128 @@ sub handler {
 		# Save new account
 		if ( $params->{pref_password} ) {
-			$params->{pref_password} = md5_hex( $params->{pref_password} );
+			# Last.fm Password (MD5)
+			if ( !($params->{pref_api_type} && $params->{pref_api_type} eq 'listenbrainz') ) {
+				$params->{pref_password} = md5_hex( $params->{pref_password} );
+			}
```
The comment seems to contradict the code: is this for LastFM or for ListenBrainz?

vysmaty replied:
It was negated. I didn't know migration and I was afraid. Edited.

michaelherger commented on Slim/Plugin/AudioScrobbler/strings.txt:
```diff
@@ -63,84 +64,148 @@ SETUP_PLUGIN_AUDIOSCROBBLER_ACCOUNTS_DELETE
+	CS	Zadejte vaše přihlašovací jméno a heslo k Last.fm či ListenBrainz.
+	DA	Skriv dit brugernavn og adgangskoden til Last.fm eller ListenBrainz.
+	DE	Geben Sie Benutzername und Kennwort für Last.fm oder ListenBrainz ein.
+	EN	Enter your Last.fm username and password or ListenBrainz.
```
Something's missing here: `... or ListenBrainz details`, `... or ListenBrainz credentials`?

michaelherger commented:
> Hopefully done responsibly and clearly on my part, but for example your link to [`public`/9.1/Slim/Plugin/DateTime/Plugin.pm#L20-L39](https://github.com/LMS-Community/slimserver/blob/public/9.1/Slim/Plugin/DateTime/Plugin.pm?rgh-link-date=2025-08-18T21%3A04%3A33Z#L20-L39) I understand, I'm interested in, but wheres is stored information about the version from which it will be migrated is not very clear me. And I don't want to break anything unnecessarily in my setup by trial and error. 😇

The base version is 0 - that's the current version unless there's some migration code in the file (which isn't for this plugin). The example should bump it to 1. You'd find the version in `<prefs folder>/plugin/audioscrobbler.pref`. I'd recommend you make a backup of that file so you can go back to before to test the migration path.

vysmaty committed:
Something's missing here: ListenBrainz credentials

Signed-off-by: Martin Šebek <10959584+vysmaty@users.noreply.github.com>

vysmaty committed:
Remove the product names

Signed-off-by: Martin Šebek <10959584+vysmaty@users.noreply.github.com>

vysmaty committed:
Wrap each label/input combination into a single span

Signed-off-by: Martin Šebek <10959584+vysmaty@users.noreply.github.com>

vysmaty committed:
Tabs rather than spaces for indentation

Signed-off-by: Martin Šebek <10959584+vysmaty@users.noreply.github.com>

vysmaty committed:
Migration $prefs from version 0 to 1: add api_type and api_url to existing accounts

Signed-off-by: Martin Šebek <10959584+vysmaty@users.noreply.github.com>

vysmaty committed:
Is this for LastFM or for ListenBrainz?

Signed-off-by: Martin Šebek <10959584+vysmaty@users.noreply.github.com>

vysmaty commented on Slim/Plugin/AudioScrobbler/Settings.pm:
```diff
@@ -61,68 +61,128 @@ sub handler {
+				return;
+			}
+			if ( $params->{pref_api_type} && $params->{pref_api_type} eq 'listenbrainz' ) {
+				# Listenbrainz
```
TODO: Implement check for a correct and functional connection to the ListenBrainz API when adding an account. And fix translations and messages.

vysmaty replied:
I've added validation to check for connections to the ListenBrainz API when adding accounts. See commit ee04934

vysmaty committed:
feat: Add ListenBrainz API connection validation on account add

This commit introduces a mechanism to verify the functionality and correctness of the ListenBrainz API connection immediately when a user attempts to add an account. Previously, no such check was performed, leading to potential issues where invalid credentials or connection problems were only discovered much later.

This ensures that only working configurations can be saved, improving user experience and preventing silent failures in scrobbling.

Signed-off-by: Martin Šebek <10959584+vysmaty@users.noreply.github.com>

vysmaty committed:
Remove fallbacks to Lastfm

Signed-off-by: Martin Šebek <10959584+vysmaty@users.noreply.github.com>

vysmaty commented:
> The base version is 0 - that's the current version unless there's some migration code in the file (which isn't for this plugin). The example should bump it to 1. You'd find the version in `<prefs folder>/plugin/audioscrobbler.pref`. I'd recommend you make a backup of that file so you can go back to before to test the migration path.

Thank you for the explanation. I've worked it out and removed the fallback. I have tested the modifications continuously. When we will be at the end, I will try the functionality with all accounts again.

michaelherger reviewed (approved):
Unfortunately I'm down sick today... But let's merge this so I can push what I've cobbled together yesterday. 

Please test my (coming) changes thoroughly. I only run basic validation with a Maloja instance I fired up quickly... I have no experience with Listenbrainz other than that!

Thanks a ton!

michaelherger commented on Slim/Plugin/AudioScrobbler/Plugin.pm:
```diff
@@ -909,6 +1277,153 @@ sub submitScrobble {
+			listened_at => $listened_at,
+		};
+	}
+	# Submit items one by one to avoid 400 errors with multiple items
+	_submitListenBrainzItems($client, \@listens, $api_url, $api_key, $cb, 0);
```
Did you try to import multiple items using `listen_type => 'import'` instead of `single`? The latter only accepts a single value, import allows you to import multiple values. 

But in reality it probably doesn't matter much, as we usually report one by one anyway. Only in case the server or network was down we'd queue up tracks.

vysmaty replied:
I haven't tried it. I saw this option, but I figured it only applies to 2-3 requests in case the server restarts and I wasn't sure if all endpoint types have this option implemented correctly.

jbylsma commented:
This is very exciting to see :) I _literally_ re-setup my IDE last night to start investigating the current Last.FM API with Lyrion when I stumbled across this PR.

@michaelherger, I'd be happy to investigate modularizing the code if you'd like to concentrate on other things, I'm just limited on time and can't give any guarantees on delivery.

I have also been interested in restoring Last.FM's "Love" feature, which had previously been "restored" in bc60fe2b for a third party implementation. With these changes, do you feel that this functionality should still be implemented externally? As this abstracts away from Last.FM, it does seem like it should be externalized, but it also seems like it could be overboard for a small bit of functionality, especially if Last.FM becomes a module within the Audioscrobbler plugin. I'm only familiar with Last.FM's capabilities, so the functionality might be limited across the ecosystems, too.

vysmaty commented:
> Unfortunately I'm down sick today... But let's merge this so I can push what I've cobbled together yesterday.
> 
> Please test my (coming) changes thoroughly. I only run basic validation with a Maloja instance I fired up quickly... I have no experience with Listenbrainz other than that!
> 
> Thanks a ton!

Thanks for working it in!

Get well soon. I'll try to test it thoroughly over time. 

* So far I've come across the need to edit the example url pattern for Multi-scrobbler from: https://scrobble.yourdomain.com/1/submit-listens to https://scrobble.yourdomain.com
Because you've simplified the url normalization. :-)

* And the need to add back **UpdateNowPlaying** with timer because the songs are Orphaned after a minute. I'm just saying that in passing. I haven't dived into the new modules to see how to work it in back.

vysmaty commented:
* It's piling up in our queue and sends repeatedly..: (Multi-scrobbler filters this as duplicates, but it will need to be cleaned.)

Don't let me disturb you. I'll just put it on the record here.

```
[25-08-21 18:53:56.7902] Slim::Plugin::AudioScrobbler::Plugin::checkScrobble (519) Palm of My Hand - Queueing track for scrobbling in 114 seconds
[25-08-21 18:53:56.7931] Slim::Plugin::AudioScrobbler::Plugin::checkScrobble (543) Queue is now: [
  {
    _url => "radioparadise://4.flac",
    a => "The%20Decemberists",
    b => "I'll%20Be%20Your%20Girl",
    i => "1755793848",
    l => 233,
    "m" => "",
    n => "",
    o => "R",
    r => "",
    t => "Severed",
  },
  {
    _url => "radioparadise://4.flac",
    a => "The%20Stone%20Roses",
    b => "Second%20Coming",
    i => "1755794081",
    l => 171,
    "m" => "",
    n => "",
    o => "R",
    r => "",
    t => "Your%20Star%20Will%20Shine",
  },
  {
    _url => "radioparadise://4.flac",
    a => "Caribombo",
    b => "Camaleon",
    i => "1755794254",
    l => 215,
    "m" => "",
    n => "",
    o => "R",
    r => "",
    t => "Sharango",
  },
  {
    _url => "radioparadise://4.flac",
    a => "The%20New%20Pornographers",
    b => "Together",
    i => "1755794469",
    l => 228,
    "m" => "",
    n => "",
    o => "R",
    r => "",
    t => "Moves",
  },
  {
    _url => "radioparadise://4.flac",
    a => "Chicane",
    b => "Everything%20we%20had%20to%20leave%20behind",
    i => "1755794698",
    l => 250,
    "m" => "",
    n => "",
    o => "R",
    r => "",
    t => "1000%20More%20Suns",
  },
  {
    _url => "radioparadise://4.flac",
    a => "Judy%20Garland",
    b => "Chasing%20Rainbows",
    i => "1755794950",
    l => 165,
    "m" => "",
    n => "",
    o => "R",
    r => "",
    t => "Somewhere%20Over%20the%20Rainbow",
  },
  {
    _url => "radioparadise://4.flac",
    a => "May%20Erlewine%20%26%20The%20Woody%20Goss%20Band",
    b => "Anyway",
    i => "1755795117",
    l => 228,
    "m" => "",
    n => "",
    o => "R",
    r => "",
    t => "Palm%20of%20My%20Hand",
  },
]
[25-08-21 18:55:45.5020] Slim::Plugin::AudioScrobbler::Plugin::newsongCallback (332) Ignoring track Main Mix - FLAC Interactive, shorter than 30 seconds
[25-08-21 18:55:50.6837] Slim::Plugin::AudioScrobbler::Plugin::newsongCallback (347) Submitting scrobble queue before now playing track
[25-08-21 18:55:50.6839] Slim::Plugin::AudioScrobbler::Plugin::submitScrobble (586) Submitting to: account=multi-scrobbler
[25-08-21 18:55:50.6845] Slim::Plugin::AudioScrobbler::API::ListenBrainz::submitScrobbleItems (196) ListenBrainz: Submitting item 1 of 7: Severed
[25-08-21 18:55:50.6858] Slim::Plugin::AudioScrobbler::Plugin::newsongCallback (420) New track to scrobble: Apocalypse, will check in 140 seconds
[25-08-21 18:55:50.8336] Slim::Plugin::AudioScrobbler::API::ListenBrainz::_submitScrobbleOK (230) ListenBrainz: Item 1 submitted successfully : OK
[25-08-21 18:55:50.8358] Slim::Plugin::AudioScrobbler::API::ListenBrainz::submitScrobbleItems (196) ListenBrainz: Submitting item 2 of 7: Your Star Will Shine
[25-08-21 18:55:50.9684] Slim::Plugin::AudioScrobbler::API::ListenBrainz::_submitScrobbleOK (230) ListenBrainz: Item 2 submitted successfully : OK
[25-08-21 18:55:50.9715] Slim::Plugin::AudioScrobbler::API::ListenBrainz::submitScrobbleItems (196) ListenBrainz: Submitting item 3 of 7: Sharango
[25-08-21 18:55:51.1115] Slim::Plugin::AudioScrobbler::API::ListenBrainz::_submitScrobbleOK (230) ListenBrainz: Item 3 submitted successfully : OK
[25-08-21 18:55:51.1148] Slim::Plugin::AudioScrobbler::API::ListenBrainz::submitScrobbleItems (196) ListenBrainz: Submitting item 4 of 7: Moves
[25-08-21 18:55:51.2470] Slim::Plugin::AudioScrobbler::API::ListenBrainz::_submitScrobbleOK (230) ListenBrainz: Item 4 submitted successfully : OK
[25-08-21 18:55:51.2500] Slim::Plugin::AudioScrobbler::API::ListenBrainz::submitScrobbleItems (196) ListenBrainz: Submitting item 5 of 7: 1000 More Suns
[25-08-21 18:55:51.4486] Slim::Plugin::AudioScrobbler::API::ListenBrainz::_submitScrobbleOK (230) ListenBrainz: Item 5 submitted successfully : OK
[25-08-21 18:55:51.4523] Slim::Plugin::AudioScrobbler::API::ListenBrainz::submitScrobbleItems (196) ListenBrainz: Submitting item 6 of 7: Somewhere Over the Rainbow
[25-08-21 18:55:51.5867] Slim::Plugin::AudioScrobbler::API::ListenBrainz::_submitScrobbleOK (230) ListenBrainz: Item 6 submitted successfully : OK
[25-08-21 18:55:51.5932] Slim::Plugin::AudioScrobbler::API::ListenBrainz::submitScrobbleItems (196) ListenBrainz: Submitting item 7 of 7: Palm of My Hand
[25-08-21 18:55:51.7303] Slim::Plugin::AudioScrobbler::API::ListenBrainz::_submitScrobbleOK (230) ListenBrainz: Item 7 submitted successfully : OK
[25-08-21 18:55:52.0017] Slim::Plugin::AudioScrobbler::API::ListenBrainz::submitNowPlaying (118) Submitting Now Playing track to ListenBrainz: Apocalypse
[25-08-21 18:55:52.1318] Slim::Plugin::AudioScrobbler::API::ListenBrainz::_submitNowPlayingOK (144) ListenBrainz Now Playing track submitted successfully: OK
```

michaelherger commented:
> * It's piling up in our queue and sends repeatedly..: (Multi-scrobbler filters this as duplicates, but it will need to be cleaned.)

So the full queue is submitted repeatedly? Adding one each time? We must reset the queue when done. I thought there should be more code deduplication between the two implementations. Last.fm does that correctly.

UpdateNowPlaying shouldn't need a timer, as it's being submitted on an event handler (`newsongCallback`). Are you saying this isn't working? I didn't find a way to show the currently playing track in Maloja, but I saw that it reported it correctly.

vysmaty commented:
Yes, the queue is not being cleared. 

Neither Maloja nor Koito have Now Playing implemented. The now playing state is maintained by repeated calls. In my opinion, a suitable interval is 30 seconds. If there is no message after 60 seconds, it disappears. I didn't find information in the API documentation, but this is how it works on Multi and probably also on ListenBrainz (I tested there only at the end, when I had it working).

michaelherger commented:
> Neither Maloja nor Koito have Now Playing implemented. The now playing state is maintained by repeated calls.

But why should we try to maintain it if it's not implemented on the server side? Do you have any reference mentioning the 60s timeout?

Maloja wouldn't do anything with it: https://github.com/krateng/maloja/blob/master/maloja/apis/listenbrainz.py#L54-L55

michaelherger commented:
Ok, found something: if I understand this code correctly, then Listenbrainz would keep the data around for as long as the duration of the track which was passed:

https://github.com/metabrainz/listenbrainz-server/blob/master/listenbrainz/webserver/views/api_tools.py#L95-L100

Which would mean: no refresh needed, no timer required. I hat timers 😁.

michaelherger commented:
> * So far I've come across the need to edit the example url pattern for Multi-scrobbler from: [scrobble.yourdomain.com/1/submit-listens](https://scrobble.yourdomain.com/1/submit-listens) to [scrobble.yourdomain.com](https://scrobble.yourdomain.com)
>   Because you've simplified the url normalization. :-)

IMHO the base URL should be the host plus whatever doesn't change only. `submit-listens` is _not_ part of the base URL. It's one specific endpoint. Otherwise the `validate-token` endpoint wouldn't work. 

I've tested the code with Maloja, which expects `http://localhost:12345/apis/listenbrainz` as the base URL.

michaelherger commented:
@jbylsma please open a new issue (or a forum thread) to discuss specifics about the `love` feature. I wouldn't even remember which plugin was using it. There are more hooks for other plugins which I doubt are being used...

vysmaty commented:
> Ok, found something: if I understand this code correctly, then Listenbrainz would keep the data around for as long as the duration of the track which was passed:
> 
> https://github.com/metabrainz/listenbrainz-server/blob/master/listenbrainz/webserver/views/api_tools.py#L95-L100
> 
> Which would mean: no refresh needed, no timer required. I hat timers 😁.

You're right that it complicates things. The way I approached it, the Multi-Scrobbler has the most development and I targeted its use. Maloja is more in maintenance mode, and Koito is pretty young and the Now Playing implementation is planned for it.  The Multi-Scrobbler uses that information for status information (Unknown/Playing, Listened x s, (x %)) otherwise it writes Orphaned and deletes it after a while. I assume it also uses it for resolving priority conflicts between multiple sources. And in my opinion, it didn't hurt others.

I tried it on ListenBrainz and you're right, it lasts until the next song.

<img width="541" height="346" alt="image" src="https://github.com/user-attachments/assets/3111a4f7-62e2-4785-83bc-f581f6163f1f" />
<img width="519" height="323" alt="image" src="https://github.com/user-attachments/assets/97e67384-35b5-4ed9-9e55-900864a95d72" />
<img width="1130" height="347" alt="image" src="https://github.com/user-attachments/assets/1bc88e53-2e3d-4df5-8d28-2e87437489eb" />

https://github.com/FoxxMD/multi-scrobbler/blob/e9e59873300fbeb5ec144203cc46d2f256936be8/src/backend/sources/PlayerState/AbstractPlayerState.ts#L127-L142
https://github.com/FoxxMD/multi-scrobbler/blob/e9e59873300fbeb5ec144203cc46d2f256936be8/src/backend/sources/PlayerState/AbstractPlayerState.ts#L91-L96

vysmaty commented:
> > * So far I've come across the need to edit the example url pattern for Multi-scrobbler from: [scrobble.yourdomain.com/1/submit-listens](https://scrobble.yourdomain.com/1/submit-listens) to [scrobble.yourdomain.com](https://scrobble.yourdomain.com)
> >   Because you've simplified the url normalization. :-)
> 
> IMHO the base URL should be the host plus whatever doesn't change only. `submit-listens` is _not_ part of the base URL. It's one specific endpoint. Otherwise the `validate-token` endpoint wouldn't work.
> 
> I've tested the code with Maloja, which expects `http://localhost:12345/apis/listenbrainz` as the base URL.

👍  Totally agree with that. First of all, I wanted to make it completely gibberish-proof, whatever is entered, so try using it, removing the slashes at the end, etc. And it didn't occur to me that there is a `validate-token`. And then it got too complicated.

vysmaty commented:
* Last.fm (v1.2 API) now does not work with custom url. For example: http://localhost:42010/apis/audioscrobbler_legacy

vysmaty commented:
* I've yet to catch the isolated error v LastFM.pm:
```
[25-08-21 23:22:54.2693] Slim::Networking::IO::Select::__ANON__ (131) Error: Select task failed calling Slim::Networking::Async::HTTP::_http_read_body: Can't locate object method "client" via package "Slim::Player::SqueezePlay" at /usr/share/perl5/Slim/Plugin/AudioScrobbler/API/LastFM.pm line 274.
; fh=Slim::Networking::Async::Socket::HTTP=GLOB(0x556398045148)
```

Maybe because I was switching the Audioscrobbler. Or maybe it has origin at: https://github.com/LMS-Community/slimserver/blob/e4df7fb2ecb5919e1c7712f0b82de559d84e3214/Slim/Plugin/AudioScrobbler/API.pm#L50C2-L50C24 (Or maybe not.)

```
[25-08-21 23:22:24.8919] Slim::Plugin::AudioScrobbler::Plugin::changeAccount (258) Changed scrobble accounts with 4 queued items, removing items:
[25-08-21 23:22:24.8921] Slim::Plugin::AudioScrobbler::Plugin::changeAccount (263)   Spider Web
[25-08-21 23:22:24.8923] Slim::Plugin::AudioScrobbler::Plugin::changeAccount (263)   Shakin' Shakin' Shakes
[25-08-21 23:22:24.8925] Slim::Plugin::AudioScrobbler::Plugin::changeAccount (263)   Nobody's Fault But Mine
[25-08-21 23:22:24.8928] Slim::Plugin::AudioScrobbler::Plugin::changeAccount (273) Changing account for player cc:cc:b1:f3:06:f1 to singularita
[25-08-21 23:22:24.8931] Slim::Plugin::AudioScrobbler::Plugin::changeAccount (273) Changing account for player cc:cc:b1:f3:06:f1 to singularita
[25-08-21 23:22:54.0787] Slim::Plugin::AudioScrobbler::Plugin::newsongCallback (347) Submitting scrobble queue before now playing track
[25-08-21 23:22:54.0790] Slim::Plugin::AudioScrobbler::Plugin::submitScrobble (586) Submitting to: account=singularita
[25-08-21 23:22:54.0793] Slim::Plugin::AudioScrobbler::API::LastFM::handshake (70) Handshaking with Last.fm: http://post.audioscrobbler.com/?hs=true&p=1.2&c=ss7&v=sc9.0.2&u=singularita&t=1755811374&a=6cd56390904e8bd77789c8e25485d9e0
[25-08-21 23:22:54.0811] Slim::Plugin::AudioScrobbler::Plugin::newsongCallback (420) New track to scrobble: Traveling Light, will check in 104 seconds
[25-08-21 23:22:54.2688] Slim::Plugin::AudioScrobbler::API::LastFM::_handshakeOK (86) Handshake OK, session id: 2a566980d97bc21b2776555c91322693, np URL: http://post.audioscrobbler.com:80/np_1.2, submit URL: http://post2.audioscrobbler.com:80/protocol_1.2
[25-08-21 23:22:54.2693] Slim::Networking::IO::Select::__ANON__ (131) Error: Select task failed calling Slim::Networking::Async::HTTP::_http_read_body: Can't locate object method "client" via package "Slim::Player::SqueezePlay" at /usr/share/perl5/Slim/Plugin/AudioScrobbler/API/LastFM.pm line 274.
; fh=Slim::Networking::Async::Socket::HTTP=GLOB(0x556398045148)
[25-08-21 23:24:43.0825] Slim::Plugin::AudioScrobbler::Plugin::checkScrobble (519) Traveling Light - Queueing track for scrobbling in 104 seconds
[25-08-21 23:24:43.0836] Slim::Plugin::AudioScrobbler::Plugin::checkScrobble (543) Queue is now: [
  {

```

vysmaty commented:
* My mistake in not checking utf-8, which is not related to refactoring.

The lines https://github.com/LMS-Community/slimserver/blob/e4df7fb2ecb5919e1c7712f0b82de559d84e3214/Slim/Plugin/AudioScrobbler/API/ListenBrainz.pm#L90-L92 at ListenBrainz.pm should be:
```
		artist_name  => decode_utf8(uri_unescape($meta->{a} // '')),
		track_name   => decode_utf8(uri_unescape($meta->{t} // '')),
		release_name => decode_utf8(uri_unescape($meta->{b} // '')),
```

Plus
`use Encode;`
somewhere at begining of things.

To display the characters well.

Or cancel the escaping of characters altogether earlier if it is not needed for logging file.

vysmaty commented:
I just now noticed you made some corrections, I hope I didn't confuse with anything above unnecessarily.

michaelherger commented:
> You're right that it complicates things. The way I approached it, the Multi-Scrobbler has the most development and I targeted its use. Maloja is more in maintenance mode, and Koito is pretty young and the Now Playing implementation is planned for it. The Multi-Scrobbler uses that information for status information (Unknown/Playing, Listened x s, (x %)) otherwise it writes Orphaned and deletes it after a while. I assume it also uses it for resolving priority conflicts between multiple sources. And in my opinion, it didn't hurt others.

I didn't get "multi-scrobbler" was a product name... I thought you were using it as the category of the alternative (local) scribbling servers... got it!

> [FoxxMD/multi-scrobbler@`e9e5987`/src/backend/sources/PlayerState/AbstractPlayerState.ts#L127-L142](https://github.com/FoxxMD/multi-scrobbler/blob/e9e59873300fbeb5ec144203cc46d2f256936be8/src/backend/sources/PlayerState/AbstractPlayerState.ts#L127-L142) [FoxxMD/multi-scrobbler@`e9e5987`/src/backend/sources/PlayerState/AbstractPlayerState.ts#L91-L96](https://github.com/FoxxMD/multi-scrobbler/blob/e9e59873300fbeb5ec144203cc46d2f256936be8/src/backend/sources/PlayerState/AbstractPlayerState.ts#L91-L96)

It's my understanding from these links that the timeouts can be configured. So I'd suggest that instead of adding the burden to LMS you could simply set those values higher (eg. 5 minutes would cover any pop song, 30 minutes "anything"). Timers always come with some overhead. They have to be managed in order not to create too many of them, it's additional code etc. which most likely 99.9+% of users will never need. Your implementation of the update handler replicated quit a bit of what already is done elsewhere (check, whether the track has changed). If you really wanted you could add another type of scrobbler of class "multi-scrobbler", inheriting from ListenBrainz, but adding those timers. That way those 99.9+% of users would not bother, even if the code was buggy 😉. 

> [25-08-21 23:22:54.2693] Slim::Networking::IO::Select::__ANON__ (131) Error: Select task failed calling Slim::Networking::Async::HTTP::_http_read_body: Can't locate object method "client" via package "Slim::Player::SqueezePlay" at /usr/share/perl5/Slim/Plugin/AudioScrobbler/API/LastFM.pm line 274.
; fh=Slim::Networking::Async::Socket::HTTP=GLOB(0x556398045148)

That was the main motivation to push some fixes. Please double check whether you still see this.

Thanks for all the feedback!

vysmaty commented:
Thanks so much for your patience and sorry for that researching in Now Playing at Listenbrainz matter. 😉

> [25-08-21 23:22:54.2693] Slim::Networking::IO::Select::ANON (131) Error: Select task failed calling Slim::Networking::Async::HTTP::_http_read_body: Can't locate object method "client" via package "Slim::Player::SqueezePlay" at /usr/share/perl5/Slim/Plugin/AudioScrobbler/API/LastFM.pm line 274.
> ; fh=Slim::Networking::Async::Socket::HTTP=GLOB(0x556398045148)

> 
> That was the main motivation to push some fixes. Please double check whether you still see this.

I didn't see it again.

Conclusion:

😊 I have tested it. (✅ Last.fm, ✅ Maloja (Last.fm legacy API), ✅ Multi-Scrobbler, ✅ Maloja (ListenBrainz API), ✅ Koito, ✅ ListenBrainz.org). Everything looks fine. I have prepared a few commits with small things that are not solved from above. ;-) )

- I encountered an error in the log when choosing wrong type of api, which returns the HTML content in such a case. It's probably unnecessary to deal with it too much, but please see it for to be sure:

```
[25-08-23 11:47:40.3400] Slim::Plugin::AudioScrobbler::API::LastFM::handshake (74) Handshaking with Last.fm: https://koito.at.prostorkolem.cz/apis/listenbrainz/?hs=true&p=1.2&c=ss7&v=sc9.0.2&u=koito_listenbrainz_api&t=1755942460&a=93d82f5b0df4cc14024f815a8888fab2
[25-08-23 11:47:40.7106] Slim::Plugin::AudioScrobbler::API::LastFM::_handshakeOK (127) <!DOCTYPE html><html lang="en" style="background-color:black"><head><meta charSet="utf-8"/><meta name="viewport" content="width=device-width, initial-scale=1"/><link rel="modulepreload" href="/assets/manifest-ff6e8ee4.js"/><link rel="modulepreload" href="/assets/entry.client-BmRZHkIv.js"/><link rel="modulepreload" href="/assets/chunk-NL6KNZEE-DfxmOBHd.js"/><link rel="modulepreload" href="/assets/index-0dO883bS.js"/><link rel="modulepreload" href="/assets/root-wofbglUK.js"/><link rel="modulepreload" href="/assets/useTheme-CBUMrKm3.js"/><link rel="modulepreload" href="/assets/SearchResults-B_rHoC6u.js"/><link rel="modulepreload" href="/assets/api-B65XGjlr.js"/><link rel="modulepreload" href="/assets/AppProvider-PySFWhaf.js"/><link rel="modulepreload" href="/assets/AsyncButton-Dzd9rf7C.js"/><link rel="icon" type="image/png" href="/favicon-96x96.png" sizes="96x96"/><link rel="icon" type="image/svg+xml" href="/favicon.svg"/><link rel="shortcut icon" href="/favicon.ico"/><link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png"/><meta name="apple-mobile-web-app-title" content="Koito"/><link rel="manifest" href="/site.webmanifest"/><link rel="preconnect" href="https://fonts.googleapis.com"/><link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="anonymous"/><link rel="stylesheet" href="/assets/root-C25kHIVB.css"/><link rel="stylesheet" href="/assets/useTheme-CjcrgoC6.css"/><link rel="stylesheet" href="https://fonts.googleapis.com/css2?family=Inter:ital,opsz,wght@0,14..32,100..900;1,14..32,100..900&amp;display=swap"/></head><body class="min-h-screen"><script>window.__reactRouterContext = {"basename":"/","future":{"unstable_middleware":false,"unstable_optimizeDeps":false,"unstable_splitRouteModules":false,"unstable_subResourceIntegrity":false,"unstable_viteEnvironmentApi":false},"routeDiscovery":{"mode":"initial"},"ssr":false,"isSpaMode":true};window.__reactRouterContext.stream = new ReadableStream({start(controller){window.__reactRouterContext.streamController = controller;}}).pipeThrough(new TextEncoderStream());</script><script type="module" async="">import "/assets/manifest-ff6e8ee4.js";
import * as route0 from "/assets/root-wofbglUK.js";
  
  window.__reactRouterRouteModules = {"root":route0};

import("/assets/entry.client-BmRZHkIv.js");</script><!--$--><script>window.__reactRouterContext.streamController.enqueue("[{\"_1\":2,\"_3\":-5,\"_4\":-5},\"loaderData\",{},\"actionData\",\"errors\"]\n");</script><!--$--><script>window.__reactRouterContext.streamController.close();</script><!--/$--><!--/$--></body></html>
[25-08-23 11:47:40.7116] Slim::Plugin::AudioScrobbler::Settings::__ANON__ (123) Error saving Audioscrobbler account: "P\x{159}i ov\x{11B}\x{159}ov\xE1n\xED va\x{161}eho \xFA\x{10D}tu na Last.fm do\x{161}lo k chyb\x{11B}. Zkuste to pros\xEDm pozd\x{11B}ji. (<!DOCTYPE html><html lang=\"en\" style=\"background-color:black\"><head><meta charSet=\"utf-8\"/><meta name=\"viewport\" content=\"width=device-width, initial-scale=1\"/><link rel=\"modulepreload\" href=\"/assets/manifest-ff6e8ee4.js\"/><link rel=\"modulepreload\" href=\"/assets/entry.client-BmRZHkIv.js\"/><link rel=\"modulepreload\" href=\"/assets/chunk-NL6KNZEE-DfxmOBHd.js\"/><link rel=\"modulepreload\" href=\"/assets/index-0dO883bS.js\"/><link rel=\"modulepreload\" href=\"/assets/root-wofbglUK.js\"/><link rel=\"modulepreload\" href=\"/assets/useTheme-CBUMrKm3.js\"/><link rel=\"modulepreload\" href=\"/assets/SearchResults-B_rHoC6u.js\"/><link rel=\"modulepreload\" href=\"/assets/api-B65XGjlr.js\"/><link rel=\"modulepreload\" href=\"/assets/AppProvider-PySFWhaf.js\"/><link rel=\"modulepreload\" href=\"/assets/AsyncButton-Dzd9rf7C.js\"/><link rel=\"icon\" type=\"image/png\" href=\"/favicon-96x96.png\" sizes=\"96x96\"/><link rel=\"icon\" type=\"image/svg+xml\" href=\"/favicon.svg\"/><link rel=\"shortcut icon\" href=\"/favicon.ico\"/><link rel=\"apple-touch-icon\" sizes=\"180x180\" href=\"/apple-touch-icon.png\"/><meta name=\"apple-mobile-web-app-title\" content=\"Koito\"/><link rel=\"manifest\" href=\"/site.webmanifest\"/><link rel=\"preconnect\" href=\"https://fonts.googleapis.com\"/><link rel=\"preconnect\" href=\"https://fonts.gstatic.com\" crossorigin=\"anonymous\"/><link rel=\"stylesheet\" href=\"/assets/root-C25kHIVB.css\"/><link rel=\"stylesheet\" href=\"/assets/useTheme-CjcrgoC6.css\"/><link rel=\"stylesheet\" href=\"https://fonts.googleapis.com/css2?family=Inter:ital,opsz,wght\@0,14..32,100..900;1,14..32,100..900&amp;display=swap\"/></head><body class=\"min-h-screen\"><script>window.__reactRouterContext = {\"basename\":\"/\",\"future\":{\"unstable_middleware\":false,\"unstable_optimizeDeps\":false,\"unstable_splitRouteModules\":false,\"unstable_subResourceIntegrity\":false,\"unstable_viteEnvironmentApi\":false},\"routeDiscovery\":{\"mode\":\"initial\"},\"ssr\":false,\"isSpaMode\":true};window.__reactRouterContext.stream = new ReadableStream({start(controller){window.__reactRouterContext.streamController = controller;}}).pipeThrough(new TextEncoderStream());</script><script type=\"module\" async=\"\">import \"/assets/manifest-ff6e8ee4.js\";\nimport * as route0 from \"/assets/root-wofbglUK.js\";\n  \n  window.__reactRouterRouteModules = {\"root\":route0};\n\nimport(\"/assets/entry.client-BmRZHkIv.js\");</script><!--\$--><script>window.__reactRouterContext.streamController.enqueue(\"[{\\\"_1\\\":2,\\\"_3\\\":-5,\\\"_4\\\":-5},\\\"loaderData\\\",{},\\\"actionData\\\",\\\"errors\\\"]\\n\");</script><!--\$--><script>window.__reactRouterContext.streamController.close();</script><!--/\$--><!--/\$--></body></html>)"
```

michaelherger commented:
Maybe you should start a new issue for new problems you encounter. Can you tell me what exact input you used to get the above? And what the message given means in English 😬?

vysmaty commented:
> Can you tell me what exact input you used to get the above? And what the message given means in English 😬?

I'm sorry. I'll try to make it clearer. That's all that's left:

> 1. Last.fm (v1.2 API) now does not work with custom url. For example: http://localhost:42010/apis/audioscrobbler_legacy
> 2. Example url pattern for Multi-scrobbler from: https://scrobble.yourdomain.com/1/submit-listens to https://scrobble.yourdomain.com


3. Error message when:

When adding **ListenBrainz credential** to new account, but unintentionally Last.fm api_type is selected. --> The HTML content is logged. From` _handshakeOK`?.


The error message for the user is in English: `There was an error verifying your account with Last.fm. Please try again later.` (I didn't notice that in the log, sorry.)


I don't consider it a problem. It notifies the user that it was unable to verify **Last.fm**. I was just puzzled by the HTML at log and ` _handshakeOK`.


That's all. I didn't think it needed an issue. If it does, please let me know. Thank you.
